src/backend/dx12/src/window.rs

```diff
@@ -3,23 +3,40 @@ use core::{self, image};
 use dxguid;
 use std::collections::VecDeque;
 use std::ptr;
+#[cfg(feature = "winit")]
 use winit;
 use winapi;
 use wio::com::ComPtr;
 use {conv, native as n, Adapter, Backend, Instance, QueueFamily};
 
-use winit::os::windows::WindowExt;
+use std::os::raw::c_void;
 
 impl Instance {
-    pub fn create_surface(&self, window: &winit::Window) -> Surface {
-        let (width, height) = window.get_inner_size_pixels().unwrap();
+    pub fn create_surface_from_hwnd(&self, hwnd: *mut c_void) -> Surface {
+        let (width, height) = unsafe {
+            use winapi::RECT;
+            use user32::GetClientRect;
+            use std::mem::zeroed;
+            let mut rect: RECT = zeroed();
+            if GetClientRect(hwnd as *mut _, &mut rect as *mut RECT) == 0 {
+                panic!("GetClientRect failed");
+            }
+            ((rect.right - rect.left) as u32, (rect.bottom - rect.top) as u32)
+        };
+
         Surface {
             factory: self.factory.clone(),
-            wnd_handle: window.get_hwnd() as *mut _,
+            wnd_handle: hwnd as *mut _,
             width: width,
             height: height,
         }
     }
+
+    #[cfg(feature = "winit")]
+    pub fn create_surface(&self, window: &winit::Window) -> Surface {
```

Comment:
oh man, I wish we could have it in the `Instance` trait. I suppose GL is the only blocker: you can't add a GL context to a window that is already created

Reply:
I don’t think it is a good idea to make ‘create_surface’ be a part of ‘Instance’, because it is implementation dependent, and some platforms (ex: game consoles) do not have winit support.

Reply:
I'm imagining a scenario: porting gfx_hal to PS4 platform.
1. Create a new backend: gfx_backend_gnm
2. Winit is not supported, and therefore it does not provide default feature "winit"
3. PS4 system provides its own windowing API for surface creation, neithor HWND nor NSView.

In conclusion, surface creation is platform dependent and it is not suitable to become a method of Instance trait.

Reply:
If we make gfx_hal also has feature="winit", it will create another problem: rustc thinks gfx_hal with winit and gfx_hal without winit are different libraries, and therefore data types of one library are not compatible to the other. It will make users pay more caution on it.

Reply:
> I don’t think it is a good idea to make create_surface be a part of Instance, because it is implementation dependent, and some platforms (ex: game consoles) do not have winit support.

True, although if we do get our hands on consoles I'd prefer `winit` to be ported on them as well.

> In conclusion, surface creation is platform dependent and it is not suitable to become a method of Instance trait

One can still have platform-specific methods for surface creation, it's not conflicting with `create_surface` being hooked up to `winit`.

> If we make gfx_hal also has feature="winit", it will create another problem: rustc thinks gfx_hal with winit and gfx_hal without winit are different libraries, and therefore data types of one library are not compatible to the other. It will make users pay more caution on it.

I don't think this is correct. If you are talking about serialization, then `winit` doesn't matter, and binary compatibility is there. If you are talking about link time errors - Cargo treats features in an additive manner, so if we have `gfx_hal` being depended upon with and without `winit` feature within a single linking unit, Cargo will just link both to the one that has the feature. No conflicts there either.

Reply:
>> I don’t think it is a good idea to make create_surface be a part of Instance, because it is implementation dependent, and some platforms (ex: game consoles) do not have winit support.

> True, although if we do get our hands on consoles I'd prefer winit to be ported on them as well.

>> In conclusion, surface creation is platform dependent and it is not suitable to become a method of Instance trait

> One can still have platform-specific methods for surface creation, it's not conflicting with create_surface being hooked up to winit.

@kvark

Sure it's great to have `winit` support game console platforms to ease prototyping!

However, I'm wondering if game developers will prefer using `winit` for retail games, since **many game developers get used to prototype games on Desktop platforms**. So I'm not certain wether it is necessary to port `winit` to game console platforms.

Besides, anyone who want to access game console related code should become registered developers due to **NDA**. Thus, we have to fork some repositories to vendors' own repositories before porting, and `winit` should be forked as well.

>> If we make gfx_hal also has feature="winit", it will create another problem: rustc thinks gfx_hal with winit and gfx_hal without winit are different libraries, and therefore data types of one library are not compatible to the other. It will make users pay more caution on it.

> I don't think this is correct. If you are talking about serialization, then winit doesn't matter, and binary compatibility is there. If you are talking about link time errors - Cargo treats features in an additive manner, so if we have gfx_hal being depended upon with and without winit feature within a single linking unit, Cargo will just link both to the one that has the feature. No conflicts there either.

OK, despite feature gate problem. I got another issues we might have: forked `winit` codebase from game console vendors' repositories. Users and gfx_hal should depend on forked `winit` instead of original `winit` otherwise cargo will treat two `winit` as different libraries (and I had run into the error many times in other projects). I think the problem is minor since users should also be registered developers to access forked version libs.